### PR TITLE
SBP add order model

### PR DESCRIPTION
**Pull request by petro2061**
In this task added new functional for managements users orders. Added new controller class with endpoints for create, update and get order and order items for user. As user you can get all orders, create order from your shopping cart and get items from your order. As admin you can changed order status.

### CODE REVIEW

File: src/main/java/project/onlinebookstore/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    private final OrderService orderService;
+
+    @Operation(summary = "Create new order",
+            description = "Returns a new order for an authorized "
+                    + "user based on the user's cart")
+    @PreAuthorize("hasRole('USER')")
+    @PostMapping
+    public OrderDto createOrder(
+            Authentication authentication,
+            @Parameter(description = "Represents the customer's shipping address")
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/controller/OrderController.java]: wrong description

File: src/main/java/project/onlinebookstore/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            @Parameter(description = "Represents the order item identifier")
+            @PathVariable @Positive Long orderItemId) {
+        return orderService.getOrderItemByOrderIdAndOrderItemId(
+                getAuthenticationUserByUserId(authentication),
+                orderId,
+                orderItemId
+        );
+    }
+
+    private Long getAuthenticationUserByUserId(Authentication authentication) {
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/controller/OrderController.java]: wrong name of method, you don't get user

File: src/main/java/project/onlinebookstore/dto/order/CreateOrderRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+package project.onlinebookstore.dto.order;
+
+import jakarta.validation.constraints.NotBlank;
+
+public record CreateOrderRequestDto(
+        @NotBlank(message = "Sipping address can't be null")
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/dto/order/CreateOrderRequestDto.java]:
```suggestion
        @NotBlank(message = "Shipping address can't be null")
```

File: src/main/java/project/onlinebookstore/dto/order/CreateOrderUpdateRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package project.onlinebookstore.dto.order;
+
+import jakarta.validation.constraints.NotNull;
+import project.onlinebookstore.model.Order;
+
+@NotNull(message = "Status cannot be null")
+public record CreateOrderUpdateRequestDto(
+        Order.Status status
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/dto/order/CreateOrderUpdateRequestDto.java:8]:
```suggestion
        @NotNull(message = "Status cannot be null")
        Order.Status status
```

File: src/main/java/project/onlinebookstore/service/order/OrderService.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+import org.springframework.data.domain.Pageable;
+import project.onlinebookstore.dto.order.CreateOrderRequestDto;
+import project.onlinebookstore.dto.order.CreateOrderUpdateRequestDto;
+import project.onlinebookstore.dto.order.OrderDto;
+import project.onlinebookstore.dto.orderitem.OrderItemDto;
+
+public interface OrderService {
+    OrderDto createOrder(Long userId, CreateOrderRequestDto orderRequestDto);
+
+    List<OrderDto> getAllOrder(Long userId, Pageable pageable);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/order/OrderService.java]:
```suggestion
    List<OrderDto> getAllOrders(Long userId, Pageable pageable);
```

File: src/main/java/project/onlinebookstore/service/order/impl/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                () -> new EntityNotFoundException("Can't find user by Id: " + userId));
+
+        if (shoppingCart.getCartItems().isEmpty()) {
+            throw new OrderProcessingException("Cart items is empty for user: " + userId);
+        }
+
+        Order order = orderMapper.toModel(shoppingCart, orderRequestDto);
+        order.getOrderItems().forEach(orderItem -> orderItem.setOrder(order));
+
+        Order savedOrder = orderRepository.save(order);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/order/impl/OrderServiceImpl.java]:
```suggestion
        orderRepository.save(order);
```

File: src/main/java/project/onlinebookstore/service/order/impl/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+        shoppingCart.getCartItems().clear();
+        shoppingCartRepository.save(shoppingCart);
+
+        return orderMapper.toDto(savedOrder);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/order/impl/OrderServiceImpl.java]:
```suggestion
        return orderMapper.toDto(order);
```

File: src/main/java/project/onlinebookstore/service/order/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    @Override
+    public List<OrderDto> getAllOrder(Long userId, Pageable pageable) {
+        List<Order> byUserIdWithItems
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/order/impl/OrderServiceImpl.java]: rename variable

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Comment by petro2061** [comment]
Corrected mentor's comments on code style

File: src/main/java/project/onlinebookstore/service/order/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+        shoppingCart.getCartItems().clear();
+        shoppingCartRepository.save(shoppingCart);
+
+        return orderMapper.toDto(order);
+    }
+
+    @Override
+    public List<OrderDto> getAllOrders(Long userId, Pageable pageable) {
+        List<Order> byUserId
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/order/impl/OrderServiceImpl.java]: rename please

File: src/main/java/project/onlinebookstore/service/order/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+    private final OrderRepository orderRepository;
+    private final OrderItemRepository orderItemRepository;
+
+    @Override
+    @Transactional
+    public OrderDto createOrder(Long userId, CreateOrderRequestDto orderRequestDto) {
+
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/order/impl/OrderServiceImpl.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
left some comments

**Comment by petro2061** [comment]
Renamed variable in  getAllOrders method

**Review by Elena-Bruyako — APPROVED** [review]

**Pull request merged**